dali/test/python/test_external_source_parallel.py

```diff
@@ -119,63 +116,73 @@ def test_parallel_fork():
         parallel_pipe.start_py_workers()
     for parallel_pipe, pipe, dtype, batch_size in pipes:
         yield check_callback, parallel_pipe, pipe, epoch_size, batch_size, dtype
+        # explicitely call py_pool close as nose might still reference parallel_pipe from the yield above
+        parallel_pipe._py_pool.close()
     # test that another pipline with forking initialization fails as there is CUDA contexts already initialized
     parallel_pipe = create_pipe(callback, 'cpu', 16, py_num_workers=4,
                                 py_start_method='fork', parallel=True)
     yield raises(RuntimeError, "Cannot fork a process when there is a CUDA context already bound to the process.")(
         build_and_run_pipeline), parallel_pipe, 1
 
-@with_setup(setup_function, teardown_function)
 def test_dtypes():
     yield from check_spawn_with_callback(ExtCallback)
 
-@with_setup(setup_function, teardown_function)
 def test_random_data():
     yield from check_spawn_with_callback(ExtCallback, shapes=[(100, 40, 3), (8, 64, 64, 3)], random_data=True)
 
-@with_setup(setup_function, teardown_function)
 def test_randomly_shaped_data():
     yield from check_spawn_with_callback(ExtCallback, shapes=[(100, 40, 3), (8, 64, 64, 3)], random_data=True, random_shape=True)
 
-@with_setup(setup_function, teardown_function)
 def test_num_outputs():
     yield from check_spawn_with_callback(ExtCallbackMultipleOutputs, ExtCallbackMultipleOutputs, num_outputs=2, dtypes=[np.uint8, np.float])
 
-@with_setup(setup_function, teardown_function)
 def test_tensor_cpu():
     yield from check_spawn_with_callback(ExtCallbackTensorCPU)
 
+@with_setup(setup_function, teardown_function)
```

Comment:
How much are these changes about refactoring and how much about adding new tests for the issue this PR fixes?

Reply:
Nose test cases generated with yield are not always immediately disposed which once again causes a problem that if yield references the pipeline, the pipeline is not garbage collected immediately and check for pool begin finalized fails. Faulty check in teardown_function didn't pick that up previously.